### PR TITLE
GH-15233: [C++] Fix CopyFiles when destination is a FileSystem with background_writes

### PR DESCRIPTION
**Pull request by Tom-Newton**
### Rationale for this change
Currently it deadlocks when trying to upload more files than we have IO threads. 

### What changes are included in this PR?
1. Use CloseAsync() inside copy_one_file so that it returns futures, instead of blocking while it waits for other tasks to complete. This avoids the deadlock.
1. Fix a segfault in FileInterface::CloseAsync() by using shared_from_this() to so that it doesn't get destructed prematurely. shared_from_this() is already used in the CloseAsync() implementation on several filesystems. After change 1 this is important when downloading to the local filesystem using CopyFiles.
1. Spawn `copy_one_file` less urgently than default, so that `background_writes` are done with higher priority. Otherwise `copy_one_file` will keep buffering more data in memory without giving the `background_writes` any chance to upload the data and drop it from memory. Therefore, without this large copies would cause OOMs.
   1. The current Arrow thread pool implementation makes provision for spawning with a set priority but the priority numbers are just ignored.
   1. I made a small change to use a `std::priority_queue` to implement the priority.
   1. There is a property of the current implementation that tasks will be scheduled in the order of calls to Spawn. There are also a few test cases in `arrow-acero-hash-aggregate-test` that fail if this property is not maintained. I'm not sure what is the correct option but for now I have ensured that tasks of the same priority are run in the order they are spawned.
   1. I think this has got to be the most contentious part of the change, with the broadest possible impact if I introduce a bug. I would appreciate some input on whether this seems like a reasonable change and I'm very happy to move it to a separate PR and/or discuss further. 

### Are these changes tested?
Added some extra automated tests. I also ran some large scale tests copying between Azure and local with a directory of about 50,000 files of varying sizes totalling 160GiB.

### Are there any user-facing changes?
1. `CopyFiles` should now work reliably
2. `ThreadPool` now runs tasks in priority order

* GitHub Issue: #15233

### CODE REVIEW

File: cpp/src/arrow/filesystem/filesystem.cc
Comment (pitrou) [inline cpp/src/arrow/filesystem/filesystem.cc:659]: Note that TaskHints are ignored by the default executor implementation.

Comment (Tom-Newton) [reply cpp/src/arrow/filesystem/filesystem.cc:659]: Yes, that is why I updated `cpp/src/arrow/util/thread_pool.cc` to make use of `TaskHints`... I've been trying to get some opinion from maintainers on whether this is a reasonable thing to do.

**Review by pitrou — CHANGES_REQUESTED** [review]
This looks good on the principle.

By the way, this seems based on a rather old version of git main, can you update?

File: cpp/src/arrow/util/thread_pool.cc
Comment (Tom-Newton) [reply cpp/src/arrow/util/thread_pool.cc:471]: Hmm... it looks like `c` is protected so it doesn't seem to like it 
```
/home/tomnewton/arrow/cpp/src/arrow/util/thread_pool.cc:471:53: error: ‘std::vector<arrow::internal::{anonymous}::QueuedTask, std::allocator<arrow::internal::{anonymous}::QueuedTask> > std::priority_queue<arrow::internal::{anonymous}::QueuedTask>::c’ is protected within this context
```

File: cpp/src/arrow/filesystem/test_util.cc
Comment (Tom-Newton) [reply cpp/src/arrow/filesystem/test_util.cc]: > Also it seems that we should actually call fs->io_context().executor()->capacity()?


:+1: I've also changed the SetCapacity to be done in a similar way. Hopefully that makes sense.

File: cpp/src/arrow/filesystem/test_util.cc
Comment (Tom-Newton) [reply cpp/src/arrow/filesystem/test_util.cc]: Yeah, that's just because I don't know what I'm doing with C++ :sweat-smile:

**Review by Tom-Newton — COMMENTED** [review]
I rebased on the latest main and I think I've addressed all the comments

File: cpp/src/arrow/util/thread_pool.cc
Comment (Tom-Newton) [reply cpp/src/arrow/util/thread_pool.cc:471]: Also maybe useful to mention the error without any cast is 
```
function "arrow::internal::<unnamed>::Task::Task(const arrow::internal::<unnamed>::Task &)" (declared implicitly) cannot be referenced -- it is a deleted functionC/C++(1776)
```

Comment (pitrou) [reply cpp/src/arrow/util/thread_pool.cc:471]: Yes, I don't think `Task` can be copied, that's why we need to move-construct it.

Comment (Tom-Newton) [reply cpp/src/arrow/util/thread_pool.cc:471]: That makes sense and its the same as before when it was a `std::dequeue`. The difference with `std::priority_queue` is that it returns a `const` reference and `std::move` requires a non-const reference. The internal container is also protected, so I don't really see any alternative to using `const_cast` and as far as I understand it should be safe in this case.  

Do you think the current implementation with `const_cast` is an acceptable solution?

Comment (pitrou) [reply cpp/src/arrow/util/thread_pool.cc:471]: Yes, it's acceptable IMHO. I was just hoping there was a cleaner alternative :)

**Review by pitrou — COMMENTED** [review]
Thanks for the update! Here are some more comments.

File: cpp/src/arrow/filesystem/test_util.cc
Comment (Tom-Newton) [reply cpp/src/arrow/filesystem/test_util.cc]: Sorry, I should have spotted that one myself.

File: cpp/src/arrow/util/thread_pool_test.cc
Comment (Tom-Newton) [reply cpp/src/arrow/util/thread_pool_test.cc]: Done

**Review by pitrou — APPROVED** [review]
+1, let's wait for CI

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Comment by pitrou** [comment]
Hmm, it looks like we'll either have to fix the serial executor to also take priorities into account (if that's at all possible?), or simply disable the priority test when threading is disabled.

**Comment by Tom-Newton** [comment]
> Hmm, it looks like we'll either have to fix the serial executor to also take priorities into account (if that's at all possible?), or simply disable the priority test when threading is disabled.

I think we can implement priorities exactly the same on `SerialExecutor` as on `ThreadPool`, so I pushed a commit to do that. I feel like this should be covered by the normal CI though, without needing to build with `ARROW_ENABLE_THREADING=OFF`, so I'm going to see how easy it is to parameterise the new tests I added.

**Comment by Tom-Newton** [comment]
>I feel like this should be covered by the normal CI though, without needing to build with `ARROW_ENABLE_THREADING=OFF`, so I'm going to see how easy it is to parameterise the new tests I added.

I think it would at least require making the constructor of `SerialExecutor`, so maybe its not really desirable to add such a test?

**Comment by Tom-Newton** [comment]
@pitrou do you have any opinion on extra tests for this? If you are happy please can you trigger the crossbow tests again.

File: cpp/src/arrow/util/thread_pool.cc
Comment (pitrou) [inline cpp/src/arrow/util/thread_pool.cc:409]: Nit: naming
```suggestion
  uint64_t spawned_tasks_count_ = 0;
```

**Review by pitrou — APPROVED** [review]
+1, just a nit

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Comment by Tom-Newton** [comment]
Sorry, it looks like there are some bits that I missed when converting `SerialExecutor` to priority queue because I never tested it locally with `ARROW_ENABLE_THREADING=OFF`

**Comment by Tom-Newton** [comment]
Ok, this time I managed to test it locally `ARROW_ENABLE_THREADING=OFF` so hopefully I haven't missed anything else.

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Comment by pitrou** [comment]
Yes, I've restarted those CI builds.

File: cpp/src/arrow/util/thread_pool.cc
Comment (Tom-Newton) [reply cpp/src/arrow/util/thread_pool.cc:410]: I just realised I forgot about this and I also should have noticed the naming convention, when I originally wrote it. Pushing a correction now

**Comment by pitrou** [comment]
Ok, thanks a lot for doing this @Tom-Newton !

**Comment by Tom-Newton** [comment]
> Ok, thanks a lot for doing this @Tom-Newton !

Thanks for reviewing

**Pull request merged**
